### PR TITLE
CURATOR-164 prevent phantom services

### PR DESCRIPTION
**Pull request by madrob**
When ServiceDiscovery reregisters services after a reconnection, it's
possible to add back a service that was already deleted. This is why
we need to remove the service from the cache before deleting it from
zookeeper.

### CODE REVIEW

File: curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceDiscovery.java [hunk 1]
```diff
@@ -297,4 +304,70 @@ public void testNoServerOnStart() throws Exception
...
         }
     }
+
+    // CURATOR-164
+    @Test
+    public void testUnregisterService() throws Exception {
+        final String name = "name";
+
+        final CountDownLatch restartLatch = new CountDownLatch(1);
+        final CountDownLatch serializeLatch = new CountDownLatch(1);
```
Comment (busbey) [inline curator-x-discovery/src/test/java/org/apache/curator/x/discovery/TestServiceDiscovery.java]: you made this latch, but then used a sleep call to wait for the serialization to finish. better ot switch to using this latch.

File: curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceDiscoveryImpl.java [hunk 2]
```diff
@@ -207,9 +207,15 @@ protected void     internalRegisterService(ServiceInstance<T> service) throws Ex
         {
             try
             {
-                CreateMode      mode = (service.getServiceType() == ServiceType.DYNAMIC) ? CreateMode.EPHEMERAL : CreateMode.PERSISTENT;
-                client.create().creatingParentsIfNeeded().withMode(mode).forPath(path, bytes);
-                isDone = true;
+                synchronized(service) {
+                    // Check that the service didn't get deleted while we were reconnecting
+                    if (services.containsKey(service.getId()))
```
Comment (Randgalt) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceDiscoveryImpl.java]: Why is this needed? The test passes without this and logically it seems unnecessary.

File: curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceDiscoveryImpl.java [hunk 3]
```diff
@@ -228,15 +234,18 @@ protected void     internalRegisterService(ServiceInstance<T> service) throws Ex
...
     {
         String          path = pathForInstance(service.getName(), service.getId());
-        try
-        {
-            client.delete().forPath(path);
-        }
-        catch ( KeeperException.NoNodeException ignore )
-        {
-            // ignore
+        synchronized(service) {
```
Comment (Randgalt) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceDiscoveryImpl.java]: I don't think the sync adds anything. The services are usually copies.

Comment (madrob) [inline curator-x-discovery/src/main/java/org/apache/curator/x/discovery/details/ServiceDiscoveryImpl.java]: Imagine the following sequence:

T1: reconnect()
T2: unregister(X)
T1: for each service in cache:
T1:   register(X)
T2: services.remove(X)
T2: client.delete(X)
T1: client.create(X)

**Comment by Randgalt** [comment]
I have an alternate fix that avoids syncs. Please review: https://github.com/apache/curator/pull/73
